doc/specs/#1595 - Suggestions UI/Snippets.md

```diff
@@ -0,0 +1,526 @@
+to the user, in the context of what they're working on.
+
+The following are some examples from VsCode, Warp. These are meant to be
+illustrative of what these menus already look like in the wild:
+
+![](img/vscode-tasks-000.gif)
+
+![](img/warp-workflows-000.gif)
+
+A prototype of saving a command directly to the user's settings, then invoking
+it via the suggestions UI
+
+![](img/save-command.gif)
+
+A prototype of reading tasks from the CWD
+
+![](img/tasks-from-cwd.gif)
+
+
+## Tenets
+
+<table>
+
+<tr><td><strong>Compatibility</strong></td><td>
+
+I considered supporting YAML for local snippets (`.wt.json`), instead of JSON.
+JSON is not super friendly to command-lines - since everything's gotta be
+encapsulated as a string. Embedding tabs `\t`, newlines `\r`, escape characters,
+is fairly straightforward. However, quotes can get complicated fast in JSON,
+since they've got to be escaped too, and with many CLI utilities also having
+separate quote-parsing rules, JSON can get unwieldy quickly.
+
+However, supporting YAML directly would require us to spec out a YAML syntax for
+these files, and also find an OSS YAML parser and implement support for it. That
+would be quite a bit more expensive than JSON.
+
+</td></tr>
+
+<tr><td><strong>Accessibility</strong></td><td>
+
+Nothing particular to call out here. The Snippets pane will need to be a11y
+tested, like most of our other UI surfaces.
+
+</td></tr>
+
+<tr><td><strong>Sustainability</strong></td><td>
+
+No substantial climate impacts expected here. We're not using expensive compute
+resources for this feature, so the impact should be comparable to any other
+Terminal feature.
+
+</td></tr>
+
+<tr><td><strong>Localization</strong></td><td>
+
+I'm mildly worried here about the potential for community-driven tasks to have
+non-localized descriptions. We may need to accept a `description:{ en-us:"",
+pt-br:"", ...}`-style map of language->string descriptions. That may just need
+to be a future consideration for now.
+
+</td></tr>
+
+<tr><td><a name="security"></a><strong>Security</strong></td><td>
+
+Another reason we shouldn't support keys being able to be lazy-bound to local
+snippets: It's entirely too easy for `malicious.exe` to create a file in
+`%HomePath%` that creates a snippet for `\u003pwn-your-machine.exe\r` (or
+similar). Any app can read your settings file, and it is again too easy for that
+malicious app to set it's own action `id` to the same as some other well-meaning
+local snippet's ID which you DO have bound to a key.
+
+</td></tr>
+
+</table>
+
+### Other potential issues
+
+Something like `wt save ping 8.8.8.8 > foo.txt` isn't going to work the way
+users want. The shell is gonna get the first crack at parsing that commandline,
+and is going to try and redirect the _output of `wt`_ to `foo.txt`.
+
+If we've got a `.wt.json` in a given directory, should we be dynamically
+adding/removing other settings too? Profiles?

```

Comment:
Running down the list of things in the settings model:
- Profile: this could be cool. I'm thinking a simple Razzle profile would be nice.
- actions: done
- color schemes: eh, not really useful imo

I think profiles is worth considering, but as a future consideration. We'd have to think about discoverability here.

Reply:
oh yeah this is something that's definitely top-of-mind for me right now: https://github.com/microsoft/devhome/issues/3005

Reply:
I'm specifically calling that thought out as a future consideration, since that's too much of a shower thought for now.